<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Visualize T-cell clonal expansion via circle-packing.
  APackOfTheClones can integrate single-cell RNA sequencing (sc-RNAseq) and T-cell receptor (TCR) libraries from the 10X genomics Single Cell Immune Profiling and Cell Ranger pipeline, to produce a simple, publication-ready visualization of the clonal expansion. 
  The method was originally implemented by Murray Christian and Ben Murrell in the following immunology study:
  Ma et al. (2021) &lt;doi:10.1126/sciimmunol.abg6356&gt;">
<title>Visualization of T-Cell Clonal Expansion • APackOfTheClones</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Visualization of T-Cell Clonal Expansion">
<meta property="og:description" content="Visualize T-cell clonal expansion via circle-packing.
  APackOfTheClones can integrate single-cell RNA sequencing (sc-RNAseq) and T-cell receptor (TCR) libraries from the 10X genomics Single Cell Immune Profiling and Cell Ranger pipeline, to produce a simple, publication-ready visualization of the clonal expansion. 
  The method was originally implemented by Murray Christian and Ben Murrell in the following immunology study:
  Ma et al. (2021) &lt;doi:10.1126/sciimmunol.abg6356&gt;">
<meta property="og:image" content="https://qile0317.github.io/APackOfTheClones/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">APackOfTheClones</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="articles/web_only/Clonal_expansion_plotting.html">Clonal expansion plotting tutorial</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header">
<img src="logo.png" class="logo" alt=""><h1 id="apackoftheclones-">APackOfTheClones <a class="anchor" aria-label="anchor" href="#apackoftheclones-"></a>
</h1>
</div>
<!-- badges: start -->

<p><code>APackOfTheClones</code> is an R package to produce “ball-packing” visualizations of the clonal expansion of T-cells in a <code>Seurat</code> object, based on its T-cell-receptor (TCR) library generated from the Cell Ranger output of 10X single-cell immune profiling.</p>
<p>The concept was first implemented in a study Ma et al.<a href="#id_1">[1]</a> by Murray Christian and Ben Murrell (@Murrellb) for nasal polyp <span class="math inline"><em>T</em><sub><em>H</em></sub></span> cells. <code>APackOfTheClones</code> counts the clonotype frequencies for each seurat/umap cluster and produces a circle packing of the frequencies to intuitively represent clonal expansion. The packing for each cluster are then plotted with roughly the same coordinates as the original UMAP reduction and color. Below is an example of <code>APackOfTheClones</code> being used on a <a href="https://www.10xgenomics.com/resources/datasets/pbm-cs-from-bal-bc-mice-tcr-enrichment-from-amplified-c-dna-1-standard-3-0-0" class="external-link">public 10X genomics dataset</a>:</p>
<p><img src="reference/figures/example.png" width="100%" align="center"></p>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p><code>APackOfTheClones</code> is registered on CRAN. To install the latest stable release, simply run the following</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"APackOfTheClones"</span><span class="op">)</span></span></code></pre></div>
<p>Alternatively, to download the latest development edition, run</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://devtools.r-lib.org/" class="external-link">devtools</a></span><span class="op">)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"Qile0317/APackOfTheClones"</span><span class="op">)</span></span></code></pre></div>
<p>If you also want the vignettes to be built locally, instead run:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"Qile0317/APackOfTheClones"</span>, build_vignettes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h2>
<p>The main ball packing function <code>clonal_expansion_plot</code> in the package takes two main arguments: - a Seurat object with a UMAP reduction - a TCR library dataframe, from <code>all_contig_annotations.csv</code> generated by Cell Ranger</p>
<p>And integrates the TCR library into the seurat object metadata(creates a new temporary seurat object). It then will produce a <em>ggplot</em> of clusters of circles at roughly the same UMAP coordinates, where the circle size represents the clone size. An example to do so is shown below.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://readr.tidyverse.org" class="external-link">readr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://qile0317.github.io/APackOfTheClones/">APackOfTheClones</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># run sc-RNAseq pipeline to produce a seurat_object</span></span>
<span></span>
<span><span class="va">seurat_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">seurat_object</span><span class="op">)</span></span>
<span><span class="va">TCR_dataframe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="st">"file_location/all_contig_annotations.csv"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># produce the ball-packing plot</span></span>
<span><span class="va">ball_pack_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/clonal_expansion_plot.html">clonal_expansion_plot</a></span><span class="op">(</span><span class="va">seurat_object</span>, <span class="va">TCR_dataframe</span><span class="op">)</span></span>
<span><span class="va">ball_pack_plot</span></span></code></pre></div>
<p>Run <code><a href="reference/clonal_expansion_plot.html">?clonal_expansion_plot</a></code> for more details and customization of the resulting plot with optional arguments, and error handling. It is also highly recommended to save the plot as an svg for use in publications.</p>
<p>In addition, <code>integrate_tcr</code> can just integrate the TCR V(D)J library into a seurat object metadata for further analyses:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">TCR_dataframe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="st">"file_location/all_contig_annotations.csv"</span><span class="op">)</span></span>
<span><span class="va">integrated_seurat_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/integrate_tcr.html">integrate_tcr</a></span><span class="op">(</span><span class="va">seurat_object</span>, <span class="va">TCR_dataframe</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># or simply modify the current seurat object</span></span>
<span><span class="va">seurat_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/integrate_tcr.html">integrate_tcr</a></span><span class="op">(</span><span class="va">seurat_object</span>, <span class="va">TCR_dataframe</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># The integrated seurat object can even be just passed to the clonal expansion plot without the TCR dataframe</span></span>
<span><span class="va">ball_pack_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/clonal_expansion_plot.html">clonal_expansion_plot</a></span><span class="op">(</span><span class="va">seurat_object</span><span class="op">)</span></span>
<span><span class="va">ball_pack_plot</span></span></code></pre></div>
<p>The function also takes care of multiple rows of data for the same cell barcode. Run <code><a href="reference/integrate_tcr.html">?integrate_tcr</a></code> for more details.</p>
</div>
<div class="section level2">
<h2 id="documentation">Documentation<a class="anchor" aria-label="anchor" href="#documentation"></a>
</h2>
<p>Comprehensive documentation, vignettes, and a changelog is deployed at <a href="https://qile0317.github.io/APackOfTheClones/" class="uri">https://qile0317.github.io/APackOfTheClones/</a></p>
<p>All exported functions has function level documentation which can be called with <code>?function</code>. Additionally, as mentioned in the <code>Installation</code> section earlier, vignettes can also be built locally. Currently, there is only one vignette named <code>"Clonal_expansion_plotting"</code> which is a walkthrough on the clonal expansion visualization workflow. All vignettes can also be viewed online at <a href="https://qile0317.github.io/APackOfTheClones/articles/" class="uri">https://qile0317.github.io/APackOfTheClones/articles/</a></p>
</div>
<div class="section level2">
<h2 id="contributing">Contributing<a class="anchor" aria-label="anchor" href="#contributing"></a>
</h2>
<p><code>APackOfTheClones</code> is currently incomplete, and more features, tests, bugfixes, and documentation are to come. However, pull requests from forked branches are more than welcome as it is mostly a solo-project at the moment. For major changes, please open an issue first to discuss what you would like to change. Please also make sure to update tests as appropriate.</p>
<p>An explanation of the algorithms will eventually be made publicly available in a either a pre-print or developer vignette.</p>
</div>
<div class="section level2">
<h2 id="contact">Contact<a class="anchor" aria-label="anchor" href="#contact"></a>
</h2>
<p>Qile Yang - <a href="mailto:qile0317@gmail.com" class="email">qile0317@gmail.com</a></p>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<p><a id="1">[1]</a> Ma, J., Tibbitt, C. A., Georén, S. K., Christian, M., Murrell, B., Cardell, L. O., Bachert, C., &amp; Coquet, J. M. (2021). Single-cell analysis pinpoints distinct populations of cytotoxic CD4+ T cells and an IL-10+CD109+ TH2 cell population in nasal polyps. Science immunology, 6(62), eabg6356. <a href="https://doi.org/10.1126/sciimmunol.abg6356" class="external-link uri">https://doi.org/10.1126/sciimmunol.abg6356</a></p>
<p><a id="2">[2]</a> atakanekiz (2019) Tutorial:Integrating VDJ sequencing data with Seurat. <a href="https://www.biostars.org/p/384640/" class="external-link uri">https://www.biostars.org/p/384640/</a></p>
</div>
<div class="section level2">
<h2 id="acknowledgements">Acknowledgements<a class="anchor" aria-label="anchor" href="#acknowledgements"></a>
</h2>
<p>Thanks for Ben Murrell (@murrelb) at the Karolinska Institute for introducing the idea and giving suggestions.</p>
</div>
</div>
  </main><aside class="col-md-3"><div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing APackOfTheClones</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Qile Yang <br><small class="roles"> Maintainer, author, copyright holder </small>  </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/Qile0317/APackOfTheClones/actions/workflows/R-CMD-check.yaml" class="external-link"><img src="https://github.com/Qile0317/APackOfTheClones/actions/workflows/R-CMD-check.yaml/badge.svg" alt="R-CMD-check"></a></li>
<li><a href="https://app.codecov.io/gh/Qile0317/APackOfTheClones?branch=main" class="external-link"><img src="https://codecov.io/gh/Qile0317/APackOfTheClones/branch/main/graph/badge.svg" alt="Codecov test coverage"></a></li>
<li><a href="https://qile0317.github.io/APackOfTheClones/"><img src="https://img.shields.io/badge/docs-stable-blue.svg" alt="Documentation"></a></li>
<li><a href="https://github.com/Qile0317/APackOfTheClones/releases/latest" class="external-link"><img src="https://img.shields.io/github/release/Qile0317/APackOfTheClones.svg" alt="Latest Release"></a></li>
<li><a href="https://lifecycle.r-lib.org/articles/stages.html#experimental" class="external-link"><img src="https://img.shields.io/badge/lifecycle-experimental-orange.svg" alt="Lifecycle: experimental"></a></li>
<li><a href="https://CRAN.R-project.org/package=APackOfTheClones" class="external-link"><img src="https://www.r-pkg.org/badges/version/APackOfTheClones" alt="CRAN status"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Qile Yang.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
