---
title: "Clonal expansion plotting tutorial"
output: rmarkdown::html_vignette
description: >
  A full walkthrough of the clonal expansion visualization workflow.
vignette: >
  %\VignetteIndexEntry{Clonal_expansion_plotting}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
data: 'Compiled: `r format(Sys.Date(), "%B %d, %Y")`'
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
knitr::opts_chunk$set(echo = FALSE)
options(rmarkdown.html_vignette.check_title = FALSE)
options(repos = c(CRAN = "http://cran.rstudio.com"))

# function to install and load packages quietly
quiet_load <- function(pkg, CRAN = TRUE, dev_dir = "Qile0317/") {
  if (!require(pkg, quietly = TRUE, character.only = TRUE)) {
    if (CRAN) {
      invisible(install.packages(pkg, quiet = TRUE, verbose = FALSE, character.only = TRUE))
    }else {
      suppressWarnings(suppressMessages(
        devtools::install_github(paste(dev_dir, pkg, sep = ""))
      ))
    }
  }
  suppressPackageStartupMessages(invisible(require(pkg, character.only = TRUE)))
}

# redefine "head" to output a nicer dataframe
view_head <- function(df) {
  knitr::kable(head(df))
}

# load misc packages
quiet_load("ggplot2")
quiet_load("cowplot")
quiet_load("Seurat")

# load packages and data
quiet_load("devtools") # ugh, had to add devtools to import
quiet_load("SCIPData", CRAN = FALSE)
if(!exists("pbmc")) {data("pbmc")}
if(!exists("tcr_dataframe")) {data("tcr_dataframe")}
```

# Introduction

Single-cell RNA sequencing (scRNA-seq) and T cell receptor (TCR) sequencing are popular techniques for studying immune cell function and disease. The combined use of such data can provide insights into clonal expansion. `APackOfTheClones` provides a simple, publication-ready method to intuitively visualize clonal expansion between different cell clusters with `ggplot2`, and can be easily slotted into any analysis pipeline.

In this user vignette, we will assume basic familiarity with `Seurat` and `R`, and a `Seurat` object has already been processed through the basic pipeline until at least the UMAP reduction has been obtained. Most importantly, clonal expansion can only be analyzed if the biosample has been run through the 10X genomics [Chromium Single Cell Immune Profiling](https://www.10xgenomics.com/products/single-cell-immune-profiling) service, and the resulting data processed by [Cell Ranger](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger).

We will be showing basic usage of the main function of the package, Namely, in the integration of the 10X genomics T-cell receptor library generated by Cell Ranger into a `Seurat` object, followed by the usage of a fully customizable clonal expansion plotting method.

# Setup the Seurat object and T cell receptor library

Before anything else can be done, we need a `Seurat` object with a UMAP reduction. 

```{r,  echo=TRUE}
library(Seurat)
library(APackOfTheClones)

# Here, a Seurat object has already been loaded named `pbmc`
print(pbmc)
```

Here, we can also see its UMAP reduction plot. It is very important that it is present within the Seurat object so that the later clonal expansion visualization can be based on the coordinates for intuitiveness.
```{r}
umap_plot <- UMAPPlot(pbmc)
umap_plot
```

To analyze clonal expansion, the T-cell library (in the form of the `all_contig_annotations.csv` file) generated by Cell Ranger has to be "integrated" into the seurat object. This simply means that the cell-level receptor information is incorporated into the `@meta.data` attribute. However, before doing so, let's quickly view the file structure:
```{r}
# an "all_contig_annotations.csv" file has already been loaded here in an object named "tcr_dataframe", but users can do so via readr::read.csv(file_location_string)
view_head(tcr_dataframe)
```

This is the first 5 rows of what the file should look like directly from cell ranger. If you for some reason have some custom version of this dataframe that is different, it is highly important that there AT LEAST exists the column with the column name `barcode` (column 1) and `raw_clonotype_id` (column 17), otherwise everything else will not work as intended.

# Integration of clonotype information

To finally incorporate the TCR(T-cell receptor library) data into the seurat object, we'll use the simple `integrate_tcr` function.

```{r}
pbmc <- integrate_tcr(pbmc, tcr_dataframe, verbose = FALSE)

# the last argument `verbose` intentionally set to `FALSE` due to the presence of a progress bar which corrupts the formatting of the vignette (it defaults to the recommended `TRUE`). Otherwise, the output to the R Console would have been the following:

#> integrating TCR library into seurat object
#   |====================================| 100%
#> Percent of unique barcodes: 31 %
```

By calling `integrate_tcr`, one can now find all the information from `all_contig_annotations.csv` within the `@meta.data` attribute. Let's take a quick look on what it looks like:

```{r}
view_head(pbmc@meta.data)
```
From this snippet of the meta data, we can see that for this particular seurat object, the first 5 columns  `orig.ident, nCount_RNA, nFeature_RNA, RNA_snn_res.0.6, seurat_clusters` were originally present. What we care about is all the new columns after them. A few key things to note:
* There are a lot of `NA` values, and that is due to 1) some cells aren't T cells, 2) some T cell receptors do not end up being sequenced for many possible reasons, 3) Of the sequenced T cells in the TCR library, there are usually a lot of duplicate barcodes (repeated TCR sequencing of the same cell) as each individual cell may have several TCR types and/or need several contigs to assemble the final sequence.
* unique data about TCRs of the same cell (same barcode) are collapsed together with `__` in the order of the contigs

Now, the seurat object can be analyzed downstream with the new itegrated information however we wish. It also is now ready to have its clonotypes counted and the clonal expansion visualized.

<details>
  <summary>**do I have to integrate the the TCR library into my seurat object?**</summary>

well... no. But there is no reason not to do so and it is heavily recommended to do so. Otherwise, we are about to see in the upcoming clonal expansion plotting function how you can avoid the integration.
</details>

<details>
  <summary>**How do I get the clonal expansion per cluster information myself?**</summary>

There is the `count_clone_sizes(seurat_obj)` function for doing so. Check the documentation by running `?count_clone_sizes` for more details.
</details>

# Ball packing visualization

All one has to do to produce the visualization is by simply running the following code:
```{r}
pbmc_expansion_plot <- clonal_expansion_plot(pbmc, verbose = FALSE)
# once again, `verbose` is set to false due to numerous progress bars but it defaults to TRUE and that is very recommended
pbmc_expansion_plot
```

A few things to note about the plot:
* The clonotype counts for each seurat cluster corresponds to their position and color in the original UMAP centroids
* The most expanded clonotypes are in the center of each circle cluster with larger sizes, symbolizing increased expansion
* There are less circles than the scRNAseq cell count due to aforementioned reasons.
* On the top left, there is a somewhat covered visual legend of the relative clone sizes
* The red and green clusters have considerable visual overlap
* The returned plot is a fully customizable `ggplot` object

As `APackOfTheClones` is in a rapid-development phase, the resulting clonal expansion plot will usually not be visually satisfactory on the first run without customizations. However, there are many optional arguments and `ggplot` tricks that we are about to explore that will help us make it publication quality.

# Plot refinement and customization

*Unfinished*

## Adjusting intra-cluster spacing
The considerable overlap between clusters (due to the algorithm's attempt to fit clusters to the original UMAP coordinates) may sometimes obstruct eachother excessively.

To account for this, there are four optional arguments in `clonal_expansion plot`:
```
repulse = FALSE,
repulsion_threshold = 1,
repulsion_strength = 1,
max_repulsion_iter = 10
```
For more details on them, read the "Arguments" section in the function documentation (`?clonal_expansion_plot`). But to summarize, first, to make the circle clusters move away from eachother, `repulse` should be set to `TRUE`, and the function should be ran AGAIN. (If you feel this excessive re-running takes too long or is inefficient for your workflow in its current form, please make an issue on the github page.) 

Setting repulse to TRUE with those default parameters should probably yield something like the following:
```{r}
pbmc_expansion_plot <- clonal_expansion_plot(pbmc, repulse = TRUE, verbose = FALSE)
```

## Adjusting and customizing size legend

## Adjusting and customizing circle size and scaling

## Possible ggplot-based modifications

# Final product
```{r}
library(ggplot2)
library(cowplot)

comparison_plot <- cowplot::plot_grid(umap_plot + ggtitle("scRNA-seq UMAP"),
                        pbmc_expansion_plot + ylim(-12,17) +
                          theme(legend.position = "none") +
                          ggtitle("APackOfTheClones clonal expansion plot"),
                        labels = "AUTO")
comparison_plot
```

# Appendix
If you really want to reproduce the plots and code in this vignette, you will need to install the data used in this vignette which has been compiled in an R package `SCIPData`. You can do so with the following code
```{r, eval = FALSE}
# install the data package
library(devtools)
devtools::install_github("Qile0317/SCIPData")

# load the package data directly into R objects in memory
data("pbmc")
data("tcr_dataframe")
```
And the rest of the code can be copied and used accordingly. 