---
title: "Clonal expansion plotting tutorial"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Clonal_expansion_plotting}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
data: 'Compiled: `r format(Sys.Date(), "%B %d, %Y")`'
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Introduction

Single-cell RNA sequencing (scRNA-seq) and T cell receptor (TCR) sequencing are popular techniques for studying immune cell function and disease. The combined use of such data can provide insights into clonal expansion. `APackOfTheClones` provides a simple, publication-ready method to intuitively visualize clonal expansion between different cell clusters with `ggplot`, and can be easily slotted into any analysis pipeline.

In this user vignette, we will assume basic familiarity with `Seurat` and `R`, and a `Seurat` object has already been processed through the basic pipeline until at least the UMAP reduction has been obtained. Most importantly, clonal expansion can only be analyzed if the biosample has been run through the 10X genomics [Chromium Single Cell Immune Profiling](https://www.10xgenomics.com/products/single-cell-immune-profiling) service, and the resulting data processed by [Cell Ranger](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger).

We will be showing basic usage of the main function of the package, Namely, in the integration of the 10X genomics T-cell receptor library generated by Cell Ranger into a `Seurat` object, followed by the usage of a fully customizable clonal expansion plotting method.

# Setup the Seurat object

Before anything else can be done, we need a `Seurat` object with a UMAP reduction. 

```{r init}
library(Seurat)
library(APackOfTheClones)
library(dplyr)

# Here, a readily-analyzed .RDS of a seurat object is loaded in with a UMAP reduction
pbmc <- readRDS("../loc_tesdat/10x.rds")
umap_plot <- DimPlot(pbmc)
umap_plot
```

To analyze clonal expansion, the T-cell library (in the form of the `all_contig_annotations.csv` file) generated by Cell Ranger has to be "integrated" into the seurat object. This simply means that the cell-level receptor information is incorporated into the `@meta.data` attribute. To do so, we'll use the `integrate_tcr` function.

```{r}
library(readr)

# read in the data into a dataframe
tcr_dataframe <- read.csv("../loc_tesdat/all_contig_annotations.csv")

# integrate the data
pbmc <- integrate_tcr(pbmc, tcr_dataframe, verbose = FALSE)

# the last argument `verbose` intentionally set to `FALSE` due to the presence of a progress bar which corrupts the formatting of the vignette. Otherwise, the output to the R Console would have been the following:

#> integrating TCR library into seurat object
#   |====================================| 100%
#> Percent of unique barcodes: 31 %
```

By calling `integrate_tcr`, one can now find all the information from `all_contig_annotations.csv` within the `@meta.data` attribute. One can read the documentation (`?integrate_tcr`) to see how this data is stored and how duplicate barcodes are handled, and how you can use it in further downstream analyses (Also see `?count_clone_sizes`). Beware also that the process at the moment is not perfect and there is room for improvement. 

Now, with the data ready, we can begin the visualization of clonal expansion. 

# Ball packing visualization

All one has to do to produce the visualization is by simply running the following code:
```{r}
pbmc_expansion_plot <- clonal_expansion_plot(pbmc, verbose = FALSE)
# once again, `verbose` is set to false due to numerous progress bars but it defaults to TRUE and that is very recommended
pbmc_expansion_plot
```

A few things to note about the plot:
- The clonotype counts for each seurat cluster corresponds to their position and color in the original UMAP centroids
- The most expanded clonotypes are in the center of each circle cluste with larger sizes, symbolizing increased expansion
- There are less circles than the scRNAseq cell count partly due to some cells are a part of the same circle, but mainly because most barcodes were for one reason or another, never detected during the immune profiling procedure.
- On the top left, there is a somewhat covered visual legend of the relative clone sizes
- The red and green clusters have considerable visual overlap
- The returned plot is a fully customizable `ggplot` object

As `APackOfTheClones` is in a rapid-development phase, the resulting clonal expansion plot will usually not be visually satisfactory on the first run without customizations. However, there are many optional arguments and `ggplot` tricks that we are about to explore that will help us make it publication quality.

# Plot refinement and customization

Unfinished

```{r}
library(ggplot2)
library(cowplot)

comparison <- plot_grid(umap_plot + ggtitle("scRNA-seq UMAP"),
                        pbmc_expansion_plot + ylim(-12,17) +
                          theme(legend.position = "none") +
                          ggtitle("APackOfTheClones clonal expansion plot"),
                        labels = "AUTO")
comparison
```

# Appendix