<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>ARCHIVED: A walkthrough of APackOfTheClones v0.1.x • APackOfTheClones</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../../apple-touch-icon-60x60.png">
<script src="../../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../../deps/headroom-0.11.0/headroom.min.js"></script><script src="../../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../../deps/search-1.0.0/fuse.min.js"></script><script src="../../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../../pkgdown.js"></script><meta property="og:title" content="ARCHIVED: A walkthrough of APackOfTheClones v0.1.x">
<meta name="description" content="A full walkthrough of the clonal expansion visualization workflow.
">
<meta property="og:description" content="A full walkthrough of the clonal expansion visualization workflow.
">
<meta property="og:image" content="https://qile0317.github.io/APackOfTheClones/logo.png">
<meta name="robots" content="noindex">
<!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-TSDH5XEN3S"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-TSDH5XEN3S');
</script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../../index.html">APackOfTheClones</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="In-development version">1.2.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../../articles/APackOfTheClones-install.html">Install</a></li>
<li class="nav-item"><a class="nav-link" href="../../articles/APackOfTheClones.html">Get started</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes">
<li><a class="dropdown-item" href="../../articles/APackOfTheClones-runs.html">Storing and Fine-Tuning APackOfTheClones Runs</a></li>
    <li><a class="dropdown-item" href="../../articles/APackOfTheClones-shared.html">Working With Clonotypes Shared Across Clusters</a></li>
    <li><a class="dropdown-item" href="../../articles/APackOfTheClones-utils.html">Clonal Analysis Utilities in APackOfTheClones</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../../news/index.html">Changelog</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-archive" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Archive</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-archive">
<li><a class="dropdown-item" href="../../articles/archive/v0-main.html">v0.1.x main vignette</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Qile0317/APackOfTheClones/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../../logo.png" class="logo" alt=""><h1>ARCHIVED: A walkthrough of APackOfTheClones v0.1.x</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/Qile0317/APackOfTheClones/blob/unstable/vignettes/archive/v0-main.Rmd" class="external-link"><code>vignettes/archive/v0-main.Rmd</code></a></small>
      <div class="d-none name"><code>v0-main.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Single-cell RNA sequencing (scRNA-seq) and T cell receptor (TCR)
sequencing are popular techniques for studying immune cell function and
disease. The combined use of such data can provide insights into clonal
expansion. <code>APackOfTheClones</code> provides a simple,
publication-ready method to intuitively visualize clonal expansion
between different cell clusters with <code>ggplot2</code>, and can be
easily slotted into any analysis pipeline.</p>
<p>In this user vignette, we will assume basic familiarity with
<code>Seurat</code> and <code>R</code>, and a <code>Seurat</code> object
has already been processed through the basic pipeline until at least the
UMAP reduction has been obtained. Most importantly, clonal expansion can
only be analyzed if the biosample has been run through the 10X genomics
<a href="https://www.10xgenomics.com/products/single-cell-immune-profiling" class="external-link">Chromium
Single Cell Immune Profiling</a> service, and the resulting data
processed by <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger" class="external-link">Cell
Ranger</a>.</p>
<p>We will be showing basic usage of the main function of the package,
Namely, in the integration of the 10X genomics T-cell receptor library
generated by Cell Ranger into a <code>Seurat</code> object, followed by
the usage of a fully customizable clonal expansion plotting method.</p>
</div>
<div class="section level2">
<h2 id="setup-the-seurat-object-and-t-cell-receptor-library">Setup the Seurat object and T cell receptor library<a class="anchor" aria-label="anchor" href="#setup-the-seurat-object-and-t-cell-receptor-library"></a>
</h2>
<p>Before anything else can be done, we need a <code>Seurat</code>
object of the single cell experiment with a dimensional reduction. The
most modern and common reduction used in publications is UMAP (Uniform
Manifold Approximation and Projection) which is what this package
defaults to, and what this vignette will use. There is also a way to
base it off of the t-SNE (t-distributed stochastic neighbor embedding)
or PCA (Principle Component Analysis) reduction though choosing the
latter is probably not the best idea.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://qile0317.github.io/APackOfTheClones/">APackOfTheClones</a></span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Here, a Seurat object has already been loaded named `pbmc`</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 31053 features across 8000 samples within 1 assay </span></span>
<span><span class="co">#&gt; Active assay: RNA (31053 features, 2000 variable features)</span></span>
<span><span class="co">#&gt;  3 layers present: counts, data, scale.data</span></span>
<span><span class="co">#&gt;  2 dimensional reductions calculated: pca, umap</span></span></code></pre>
<p>Here, we can also see its UMAP reduction plot. It is very important
that it is present within the Seurat object so that the later clonal
expansion visualization can be based on the coordinates for
intuitiveness.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">umap_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">UMAPPlot</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span>
<span><span class="va">umap_plot</span></span></code></pre></div>
<p><img src="v0-main_files/figure-html/get_umap-1.png" width="700"></p>
<p>To analyze clonal expansion, the T-cell library (in the form of the
<code>all_contig_annotations.csv</code> file) generated by Cell Ranger
has to be “integrated” into the seurat object. This simply means that
the cell-level receptor information is incorporated into the
<code>@meta.data</code> attribute. However, before doing so, let’s
quickly view the file structure. An r dataframe of the read-in file has
been loaded named <code>tcr_dataframe</code>:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">tcr_dataframe</span><span class="op">)</span></span></code></pre></div>
<table style="width:100%;" class="table">
<colgroup>
<col width="7%">
<col width="3%">
<col width="11%">
<col width="6%">
<col width="2%">
<col width="2%">
<col width="3%">
<col width="2%">
<col width="3%">
<col width="2%">
<col width="4%">
<col width="4%">
<col width="6%">
<col width="18%">
<col width="2%">
<col width="1%">
<col width="6%">
<col width="9%">
</colgroup>
<thead><tr class="header">
<th align="left">barcode</th>
<th align="left">is_cell</th>
<th align="left">contig_id</th>
<th align="left">high_confidence</th>
<th align="right">length</th>
<th align="left">chain</th>
<th align="left">v_gene</th>
<th align="left">d_gene</th>
<th align="left">j_gene</th>
<th align="left">c_gene</th>
<th align="left">full_length</th>
<th align="left">productive</th>
<th align="left">cdr3</th>
<th align="left">cdr3_nt</th>
<th align="right">reads</th>
<th align="right">umis</th>
<th align="left">raw_clonotype_id</th>
<th align="left">raw_consensus_id</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">AAACCTGAGTTAACGA-1</td>
<td align="left">True</td>
<td align="left">AAACCTGAGTTAACGA-1_contig_1</td>
<td align="left">True</td>
<td align="right">362</td>
<td align="left">TRB</td>
<td align="left">None</td>
<td align="left">None</td>
<td align="left">TRBJ2-5</td>
<td align="left">TRBC1</td>
<td align="left">False</td>
<td align="left">None</td>
<td align="left">None</td>
<td align="left">None</td>
<td align="right">23679</td>
<td align="right">4</td>
<td align="left">clonotype4</td>
<td align="left">None</td>
</tr>
<tr class="even">
<td align="left">AAACCTGAGTTAACGA-1</td>
<td align="left">True</td>
<td align="left">AAACCTGAGTTAACGA-1_contig_2</td>
<td align="left">True</td>
<td align="right">872</td>
<td align="left">TRB</td>
<td align="left">TRBV19</td>
<td align="left">TRBD1</td>
<td align="left">TRBJ1-1</td>
<td align="left">TRBC1</td>
<td align="left">True</td>
<td align="left">True</td>
<td align="left">CASSIRVNTEVFF</td>
<td align="left">TGTGCCAGCAGTATCAGGGTAAACACAGAAGTCTTCTTT</td>
<td align="right">3993</td>
<td align="right">3</td>
<td align="left">clonotype4</td>
<td align="left">clonotype4_consensus_2</td>
</tr>
<tr class="odd">
<td align="left">AAACCTGAGTTAACGA-1</td>
<td align="left">True</td>
<td align="left">AAACCTGAGTTAACGA-1_contig_3</td>
<td align="left">True</td>
<td align="right">324</td>
<td align="left">Multi</td>
<td align="left">None</td>
<td align="left">None</td>
<td align="left">TRAJ50</td>
<td align="left">TRBC1</td>
<td align="left">False</td>
<td align="left">None</td>
<td align="left">None</td>
<td align="left">None</td>
<td align="right">4870</td>
<td align="right">2</td>
<td align="left">clonotype4</td>
<td align="left">None</td>
</tr>
<tr class="even">
<td align="left">AAACCTGAGTTAACGA-1</td>
<td align="left">True</td>
<td align="left">AAACCTGAGTTAACGA-1_contig_6</td>
<td align="left">True</td>
<td align="right">749</td>
<td align="left">TRA</td>
<td align="left">TRAV10D</td>
<td align="left">None</td>
<td align="left">TRAJ49</td>
<td align="left">TRAC</td>
<td align="left">True</td>
<td align="left">True</td>
<td align="left">CAARDTGYQNFYF</td>
<td align="left">TGTGCAGCAAGGGACACGGGTTACCAGAACTTCTATTTT</td>
<td align="right">14844</td>
<td align="right">5</td>
<td align="left">clonotype4</td>
<td align="left">clonotype4_consensus_1</td>
</tr>
<tr class="odd">
<td align="left">AAACCTGCATCCGGGT-1</td>
<td align="left">True</td>
<td align="left">AAACCTGCATCCGGGT-1_contig_1</td>
<td align="left">True</td>
<td align="right">557</td>
<td align="left">TRA</td>
<td align="left">TRAV19</td>
<td align="left">None</td>
<td align="left">TRAJ56</td>
<td align="left">TRAC</td>
<td align="left">True</td>
<td align="left">True</td>
<td align="left">CAAGGGNNKLTF</td>
<td align="left">TGCGCAGCAGGAGGAGGCAATAATAAGCTGACTTTT</td>
<td align="right">14438</td>
<td align="right">4</td>
<td align="left">clonotype5</td>
<td align="left">clonotype5_consensus_2</td>
</tr>
<tr class="even">
<td align="left">AAACCTGCATCCGGGT-1</td>
<td align="left">True</td>
<td align="left">AAACCTGCATCCGGGT-1_contig_2</td>
<td align="left">True</td>
<td align="right">619</td>
<td align="left">TRB</td>
<td align="left">TRBV5</td>
<td align="left">TRBD1</td>
<td align="left">TRBJ1-4</td>
<td align="left">TRBC1</td>
<td align="left">True</td>
<td align="left">True</td>
<td align="left">CASSLTGISNERLFF</td>
<td align="left">TGTGCCAGCAGCCTCACAGGGATTTCCAACGAAAGATTATTTTTC</td>
<td align="right">3721</td>
<td align="right">2</td>
<td align="left">clonotype5</td>
<td align="left">clonotype5_consensus_1</td>
</tr>
</tbody>
</table>
<p>This is the first 5 rows of what the file should look like directly
from cell ranger. If you for some reason have some custom version of
this dataframe that is different, it is highly important that there AT
LEAST exists the column with the column name <code>barcode</code>
(column 1) and <code>raw_clonotype_id</code> (column 17), otherwise
everything else will not work as intended.</p>
<details><summary><strong>how do I load in the raw file?</strong>
</summary><div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://readr.tidyverse.org" class="external-link">readr</a></span><span class="op">)</span></span>
<span><span class="va">tcr_dataframe</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu">read_csv</span><span class="op">(</span><span class="st">"your_file_location/all_contig_annotations.csv"</span><span class="op">)</span></span></code></pre></div>
</details>
</div>
<div class="section level2">
<h2 id="integration-of-clonotype-information">Integration of clonotype information<a class="anchor" aria-label="anchor" href="#integration-of-clonotype-information"></a>
</h2>
<p>To finally incorporate the TCR(T-cell receptor library) data into the
seurat object, we’ll use the simple <code>integrate_tcr</code>
function.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/integrate_tcr.html">integrate_tcr</a></span><span class="op">(</span><span class="va">pbmc</span>, <span class="va">tcr_dataframe</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>Note that the last argument <code>verbose</code> was intentionally
set to <code>FALSE</code> due to the presence of a progress bar which
corrupts the formatting of the vignette (it defaults to the recommended
<code>TRUE</code>). Otherwise, the output to the R Console would have
been the following:</p>
<pre><code><span><span class="co">#&gt; integrating TCR library into seurat object</span></span>
<span><span class="co">#&gt;   |====================================| 100%</span></span>
<span><span class="co">#&gt; Percent of unique barcodes: 31 %</span></span></code></pre>
<p>By calling <code>integrate_tcr</code>, one can now find all the
information from <code>all_contig_annotations.csv</code> within the
<code>@meta.data</code> attribute. Let’s take a quick look on what it
looks like:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="3%">
<col width="2%">
<col width="1%">
<col width="2%">
<col width="2%">
<col width="2%">
<col width="3%">
<col width="1%">
<col width="18%">
<col width="2%">
<col width="3%">
<col width="2%">
<col width="3%">
<col width="1%">
<col width="5%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="5%">
<col width="14%">
<col width="4%">
<col width="1%">
<col width="2%">
<col width="8%">
</colgroup>
<thead><tr class="header">
<th align="left"></th>
<th align="left">orig.ident</th>
<th align="right">nCount_RNA</th>
<th align="right">nFeature_RNA</th>
<th align="left">RNA_snn_res.0.6</th>
<th align="left">seurat_clusters</th>
<th align="left">barcode</th>
<th align="left">is_cell</th>
<th align="left">contig_id</th>
<th align="left">high_confidence</th>
<th align="left">length</th>
<th align="left">chain</th>
<th align="left">v_gene</th>
<th align="left">d_gene</th>
<th align="left">j_gene</th>
<th align="left">c_gene</th>
<th align="left">full_length</th>
<th align="left">productive</th>
<th align="left">cdr3</th>
<th align="left">cdr3_nt</th>
<th align="left">reads</th>
<th align="left">umis</th>
<th align="left">raw_clonotype_id</th>
<th align="left">raw_consensus_id</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">AAACCTGAGCAACGGT-1</td>
<td align="left">SeuratProject</td>
<td align="right">3456</td>
<td align="right">1203</td>
<td align="left">1</td>
<td align="left">1</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
</tr>
<tr class="even">
<td align="left">AAACCTGAGCAGCCTC-1</td>
<td align="left">SeuratProject</td>
<td align="right">7335</td>
<td align="right">2334</td>
<td align="left">8</td>
<td align="left">8</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
</tr>
<tr class="odd">
<td align="left">AAACCTGAGTTAACGA-1</td>
<td align="left">SeuratProject</td>
<td align="right">2949</td>
<td align="right">1144</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">AAACCTGAGTTAACGA-1</td>
<td align="left">True</td>
<td align="left">AAACCTGAGTTAACGA-1_contig_1__AAACCTGAGTTAACGA-1_contig_2__AAACCTGAGTTAACGA-1_contig_3__AAACCTGAGTTAACGA-1_contig_6</td>
<td align="left">True</td>
<td align="left">324__362__749__872</td>
<td align="left">Multi__TRA__TRB</td>
<td align="left">None__TRAV10D__TRBV19</td>
<td align="left">None__TRBD1</td>
<td align="left">TRAJ49__TRAJ50__TRBJ1-1__TRBJ2-5</td>
<td align="left">TRAC__TRBC1</td>
<td align="left">False__True</td>
<td align="left">None__True</td>
<td align="left">CAARDTGYQNFYF__CASSIRVNTEVFF__None</td>
<td align="left">None__TGTGCAGCAAGGGACACGGGTTACCAGAACTTCTATTTT__TGTGCCAGCAGTATCAGGGTAAACACAGAAGTCTTCTTT</td>
<td align="left">3993__4870__14844__23679</td>
<td align="left">2__3__4__5</td>
<td align="left">clonotype4</td>
<td align="left">clonotype4_consensus_1__clonotype4_consensus_2__None</td>
</tr>
<tr class="even">
<td align="left">AAACCTGCATCCGGGT-1</td>
<td align="left">SeuratProject</td>
<td align="right">3586</td>
<td align="right">1017</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">AAACCTGCATCCGGGT-1</td>
<td align="left">True</td>
<td align="left">AAACCTGCATCCGGGT-1_contig_1__AAACCTGCATCCGGGT-1_contig_2</td>
<td align="left">True</td>
<td align="left">557__619</td>
<td align="left">TRA__TRB</td>
<td align="left">TRAV19__TRBV5</td>
<td align="left">None__TRBD1</td>
<td align="left">TRAJ56__TRBJ1-4</td>
<td align="left">TRAC__TRBC1</td>
<td align="left">True</td>
<td align="left">True</td>
<td align="left">CAAGGGNNKLTF__CASSLTGISNERLFF</td>
<td align="left">TGCGCAGCAGGAGGAGGCAATAATAAGCTGACTTTT__TGTGCCAGCAGCCTCACAGGGATTTCCAACGAAAGATTATTTTTC</td>
<td align="left">3721__14438</td>
<td align="left">2__4</td>
<td align="left">clonotype5</td>
<td align="left">clonotype5_consensus_1__clonotype5_consensus_2</td>
</tr>
<tr class="odd">
<td align="left">AAACCTGCATCTACGA-1</td>
<td align="left">SeuratProject</td>
<td align="right">2659</td>
<td align="right">1141</td>
<td align="left">3</td>
<td align="left">3</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
</tr>
<tr class="even">
<td align="left">AAACCTGGTCTAACGT-1</td>
<td align="left">SeuratProject</td>
<td align="right">5266</td>
<td align="right">1189</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">AAACCTGGTCTAACGT-1</td>
<td align="left">True</td>
<td align="left">AAACCTGGTCTAACGT-1_contig_1__AAACCTGGTCTAACGT-1_contig_2</td>
<td align="left">True</td>
<td align="left">565__651</td>
<td align="left">TRA__TRB</td>
<td align="left">TRAV13D-2__TRBV13-3</td>
<td align="left">None__TRBD1</td>
<td align="left">TRAJ38__TRBJ2-7</td>
<td align="left">TRAC__TRBC1</td>
<td align="left">True</td>
<td align="left">True</td>
<td align="left">CAIDPPNVGDNSKLIW__CASSDDRVGEQYF</td>
<td align="left">TGTGCCAGCAGTGACGACAGGGTGGGGGAACAGTACTTC__TGTGCTATAGACCCCCCTAATGTTGGTGACAACAGTAAGCTGATTTGG</td>
<td align="left">7172__13349</td>
<td align="left">2__8</td>
<td align="left">clonotype6</td>
<td align="left">clonotype6_consensus_1__clonotype6_consensus_2</td>
</tr>
</tbody>
</table>
<p>From this snippet of the meta data, we can see that for this
particular seurat object, the first 5 columns
<code>orig.ident, nCount_RNA, nFeature_RNA, RNA_snn_res.0.6, seurat_clusters</code>
were originally present. What we care about is all the new columns after
them. A few key things to note: * There are a lot of <code>NA</code>
values, and that is due to 1) some cells aren’t T cells, 2) some T cell
receptors do not end up being sequenced for many possible reasons, 3) Of
the sequenced T cells in the TCR library, there are usually a lot of
duplicate barcodes (repeated TCR sequencing of the same cell) as each
individual cell may have several TCR types and/or need several contigs
to assemble the final sequence. * unique data about TCRs of the same
cell (same barcode) are collapsed together with <code>__</code> in the
order of the contigs</p>
<p>Now, the seurat object can be analyzed downstream with the new
integrated information however we wish. It also is now ready to have its
clonotypes counted and the clonal expansion visualized.</p>
<details><summary><strong>do I HAVE to integrate the the TCR library into my seurat
object?</strong>
</summary>
well… no. But there is no reason not to do so and it is heavily
recommended to do so. Otherwise, we are about to see in the upcoming
clonal expansion plotting function how you can avoid the integration.
</details><details><summary><strong>How do I get the clonal expansion per cluster information
myself?</strong>
</summary>
There is the <code>count_clone_sizes(seurat_obj)</code> function for
doing so. Check the documentation by running
<code><a href="../../reference/count_clone_sizes.html">?count_clone_sizes</a></code> for more details.
</details>
</div>
<div class="section level2">
<h2 id="ball-packing-visualization">Ball packing visualization<a class="anchor" aria-label="anchor" href="#ball-packing-visualization"></a>
</h2>
<p>All one has to do to produce the visualization is by simply running
the following code:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_expansion_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/clonal_expansion_plot.html">clonal_expansion_plot</a></span><span class="op">(</span><span class="va">pbmc</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co"># once again, `verbose` is set to false but it defaults to TRUE</span></span>
<span><span class="va">pbmc_expansion_plot</span></span></code></pre></div>
<p><img src="v0-main_files/figure-html/main_plt-1.png" width="700"></p>
<details><summary><strong>How do I base the clonal expansion plot off of the t-SNE or PCA
reduction?</strong>
</summary><p><code>clonal_expansion_plot</code> has an optional argument
<code>reduction</code> which defaults to <code>'umap'</code> that
handles which reduction the plot is based off of. The argument can be
set to <code>'tsne'</code> or <code>'pca'</code> like so:</p>
<pre><code><span><span class="co"># t-SNE based plot</span></span>
<span><span class="va">pbmc_expansion_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/clonal_expansion_plot.html">clonal_expansion_plot</a></span><span class="op">(</span><span class="va">pbmc</span>, reduction <span class="op">=</span> <span class="st">'tsne'</span><span class="op">)</span></span>
<span><span class="co"># PCA based plot</span></span>
<span><span class="va">pbmc_expansion_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/clonal_expansion_plot.html">clonal_expansion_plot</a></span><span class="op">(</span><span class="va">pbmc</span>, reduction <span class="op">=</span> <span class="st">'pca'</span><span class="op">)</span></span></code></pre>
</details><div class="section level3">
<h3 id="a-few-things-to-note-about-the-plot">A few things to note about the plot<a class="anchor" aria-label="anchor" href="#a-few-things-to-note-about-the-plot"></a>
</h3>
<ul>
<li>The clonotype counts for each seurat cluster corresponds to their
position and color in the original UMAP centroids</li>
<li>The most expanded clonotypes are in the center of each circle
cluster with larger sizes, symbolizing increased expansion</li>
<li>There are less circles than the scRNAseq cell count due to
aforementioned reasons.</li>
<li>On the top left, there is a somewhat covered visual legend of the
relative clone sizes</li>
<li>The red and green clusters have considerable visual overlap</li>
<li>The returned plot is a fully customizable <code>ggplot</code>
object</li>
</ul>
<p>As <code>APackOfTheClones</code> is in a rapid-development phase, the
resulting clonal expansion plot will usually not be visually
satisfactory on the first run without customizations. However, there are
many optional arguments and <code>ggplot</code> tricks that we are about
to explore that will help us make it publication quality.</p>
<details><summary><strong>How do I create the plot without integrating the Seurat object
and TCR library?</strong>
</summary><p><code>clonal_expansion_plot</code> will also work if the user
provides the original <code>seurat</code> object and the raw TCR library
dataframe, like so:</p>
<pre><code><span><span class="co"># let pbmc be the raw seurat object before the integration</span></span>
<span><span class="va">pbmc_expansion_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/clonal_expansion_plot.html">clonal_expansion_plot</a></span><span class="op">(</span><span class="va">pbmc</span>, <span class="va">tcr_dataframe</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre>
</details>
</div>
<div class="section level3">
<h3 id="adjusting-intra-cluster-spacing">Adjusting intra-cluster spacing<a class="anchor" aria-label="anchor" href="#adjusting-intra-cluster-spacing"></a>
</h3>
<p>The considerable overlap between clusters (due to the algorithm’s
attempt to fit clusters to the original UMAP coordinates) may sometimes
obstruct eachother excessively.</p>
<p>To account for this, there are four optional arguments in
<code>clonal_expansion plot</code>:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a>repulse <span class="ot">=</span> <span class="cn">FALSE</span>,</span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a>repulsion_threshold <span class="ot">=</span> <span class="dv">1</span>,</span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a>repulsion_strength <span class="ot">=</span> <span class="dv">1</span>,</span>
<span id="cb12-4"><a href="#cb12-4" tabindex="-1"></a>max_repulsion_iter <span class="ot">=</span> <span class="dv">10</span></span></code></pre></div>
<p>For more details on them, read the “Arguments” section in the
function documentation (<code><a href="../../reference/clonal_expansion_plot.html">?clonal_expansion_plot</a></code>). But to
summarize, first, to make the circle clusters move away from eachother,
<code>repulse</code> should be set to <code>TRUE</code>, and the
function should be ran AGAIN. (If you feel this excessive re-running
takes too long or is inefficient for your workflow in its current form,
please make an issue on the github page.)</p>
<p>Setting repulse to TRUE with those default parameters should probably
yield something like the following:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_expansion_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/clonal_expansion_plot.html">clonal_expansion_plot</a></span><span class="op">(</span><span class="va">pbmc</span>, repulse <span class="op">=</span> <span class="cn">TRUE</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">pbmc_expansion_plot</span></span></code></pre></div>
<p><img src="v0-main_files/figure-html/repulse-1.png" width="700"></p>
<p>As we can see, the red and green clusters have “moved away” from each
other, and there is very little overlap remaining. All other clusters
which were not touching remain in their original positions.</p>
<details><summary><strong>Do I HAVE to re-run <code>clonal_expansion_plot</code> each time
I want to modify a parameter?</strong>
</summary>
Unfortunately, yes. With the current version (v1) of
<code>APackOfTheClones</code>, to change the appearance of the plot,
<code>clonal_expansion_plot</code> has to be re-ran with modified
parameters each time, which unfortunately may be slow/tedious if a lot
of modifications are required. Please create a github issue or contact
the author if a new version should be made to prevent this.
</details><p>If you are still unhappy with the spacing of the plot, see the
following ways to adjust the spacing:</p>
<ul>
<li>
<code>repulsion_threshold</code> indicates the amount of
<code>ggplot2</code> units of overlap between clusters that are
acceptable. It defaults to <code>1</code>, meaning that two clusters
that overlap by about 1 unit are considered by the repulsion algorithm
to not be overlapping. Increasing this number will increase the amount
of overlap between clusters, and decreasing this number will do the
opposite, while decreasing it to negative values will incur additional
spacing between clusters. However, using negative spacing may cause the
plot to look very spaced out. Keep reading to see alternatives to doing
so.</li>
<li>
<code>repulsion_strength</code> relates to how much the clusters
should repel each other. The repulsion algorithm works in iterations,
where for each iteration, each cluster “pushes” each other away from
eachother by some amount. Increasing this value will cause extra
“pushing” during each iteration. However, increasing this factor too
much may result once again in a very visually unpleasant plot. Note that
closer clusters will repell eachother more, and larger clusters will
repel smaller clusters more.</li>
<li>
<code>max_repulsion_iter</code> indicates the number of iterations
where clusters should repel eachother. Increasing this number would
ensure that clusters will (almost always) for sure not be overlapping. A
trick with this parameter to make more pleasant plots is to decrease
<code>repulsion_strength</code> and increase
<code>max_repulsion_iter</code> to possibly make a more pleasant
arrangement of clusters.</li>
</ul>
</div>
<div class="section level3">
<h3 id="adjusting-and-customizing-size-legend">Adjusting and customizing size legend<a class="anchor" aria-label="anchor" href="#adjusting-and-customizing-size-legend"></a>
</h3>
<p>The plot we’ve generated here has a size legend on the top left, but
we see that it is partially outside of the plot. Unfortunately the
algorithm for placing the legend is far from perfect in version 0 so the
user will have to do a little extra work.</p>
<p>There are the following six parameters to adjust the legend:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a>add_size_legend <span class="ot">=</span> <span class="cn">TRUE</span>,</span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a>legend_sizes <span class="ot">=</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">5</span>, <span class="dv">50</span>),</span>
<span id="cb14-3"><a href="#cb14-3" tabindex="-1"></a>legend_position <span class="ot">=</span> <span class="st">"top_left"</span>,</span>
<span id="cb14-4"><a href="#cb14-4" tabindex="-1"></a>legend_buffer <span class="ot">=</span> <span class="fl">1.5</span>,</span>
<span id="cb14-5"><a href="#cb14-5" tabindex="-1"></a>legend_color <span class="ot">=</span> <span class="st">"#808080"</span>,</span>
<span id="cb14-6"><a href="#cb14-6" tabindex="-1"></a>legend_spacing <span class="ot">=</span> <span class="fl">0.4</span></span></code></pre></div>
<p>Here, we can adjust the argument <code>legend_buffer</code>, which
indicates the amount of <code>ggplot2</code> units to move the legend
“inwards” by. We can also slightly increase <code>legend_spacing</code>,
an argument specifying the vertical distance between circles in the
legend.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># the previous arguments have to be maintained!</span></span>
<span><span class="va">pbmc_expansion_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/clonal_expansion_plot.html">clonal_expansion_plot</a></span><span class="op">(</span></span>
<span>  <span class="va">pbmc</span>,</span>
<span>  legend_buffer <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  legend_spacing <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span>  repulse <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="va">pbmc_expansion_plot</span></span></code></pre></div>
<p><img src="v0-main_files/figure-html/adj_legend-1.png" width="700"></p>
<p>We can even change the circle (clone) sizes displayed in the legend
by changing the vector <code>legend_sizes</code>:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_expansion_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/clonal_expansion_plot.html">clonal_expansion_plot</a></span><span class="op">(</span></span>
<span>  <span class="va">pbmc</span>,</span>
<span>  legend_sizes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">5</span>, <span class="fl">25</span>, <span class="fl">50</span><span class="op">)</span>,</span>
<span>  legend_buffer <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  legend_spacing <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span>  repulse <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="va">pbmc_expansion_plot</span></span></code></pre></div>
<p><img src="v0-main_files/figure-html/adj_legend_sizes-1.png" width="700"></p>
<p>More details about these arguments can be read in the function
documentation.</p>
</div>
<div class="section level3">
<h3 id="adjusting-and-customizing-circle-size-and-scaling">Adjusting and customizing circle size and scaling<a class="anchor" aria-label="anchor" href="#adjusting-and-customizing-circle-size-and-scaling"></a>
</h3>
<p>So far, the sizes of the individual circles (clones) have seemed to
look quite pleasant, fitting roughly into their UMAP cluster centroids
snugly. However, this may not always be the case. A likely very common
issue upon the initialization of the plot with default arguments is that
the circles may be too small or too large.</p>
<p>In these cases, the argument <code>clone_scale_factor</code> should
be changed. The argument defaults to 0.1, which is the multiplicative
scale factor for the square root radius of a clone. So if a clone size
is four, and the factor is 0.1, then the actual radius of the circle on
the plot would be <code>sqrt(4) * 0.1</code> which equates to 0.2.</p>
<p>Additionally, <code>rad_scale_factor</code> can optionally be changed
to adjust the intra-circular spacing. The factor defaults to 0.95 and
indicates the multiplicative amount to adjust the radius of the circles
after they have been placed. decreasing this value (slightly) will
basically make each circle smaller in place, increasing the spacing
between then,</p>
<p>For demonstration sake, we’ll change the
<code>clone_scale_factor</code> to 0.07 and
<code>rad_scale_factor</code> to 0.9:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_expansion_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/clonal_expansion_plot.html">clonal_expansion_plot</a></span><span class="op">(</span></span>
<span>  <span class="va">pbmc</span>,</span>
<span>  clone_scale_factor <span class="op">=</span> <span class="fl">0.07</span>,</span>
<span>  rad_scale_factor <span class="op">=</span> <span class="fl">0.9</span>,</span>
<span>  legend_sizes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">5</span>, <span class="fl">25</span>, <span class="fl">50</span><span class="op">)</span>,</span>
<span>  legend_buffer <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  legend_spacing <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span>  repulse <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="va">pbmc_expansion_plot</span></span></code></pre></div>
<p><img src="v0-main_files/figure-html/scale_fac-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="other-possible-modifications">Other possible modifications<a class="anchor" aria-label="anchor" href="#other-possible-modifications"></a>
</h3>
<p>There are a few other arguments this vignette did not cover which can
be read in the function documentation. However, to end off, one more
important argument to consider is <code>use_default_theme = TRUE</code>.
It makes the plot have the same axis labels and general theme as the
original UMAP plot. If this isn’t to your liking or you want to
customize the plot your own way (since
<code>clonal_expansion_plot</code> generates a <code>ggplot</code>
object!), simply set it to <code>FALSE</code> and you’ll have a very
minimally themed plot:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../../reference/clonal_expansion_plot.html">clonal_expansion_plot</a></span><span class="op">(</span></span>
<span>  <span class="va">pbmc</span>,</span>
<span>  use_default_theme <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  clone_scale_factor <span class="op">=</span> <span class="fl">0.07</span>,</span>
<span>  rad_scale_factor <span class="op">=</span> <span class="fl">0.9</span>,</span>
<span>  legend_sizes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">5</span>, <span class="fl">25</span>, <span class="fl">50</span><span class="op">)</span>,</span>
<span>  legend_buffer <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  legend_spacing <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span>  repulse <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="v0-main_files/figure-html/blankplt-1.png" width="700"></p>
<p>Additionally, if you really want the plot to reflect the original
UMAP plot, you could try using <code>ggplot2</code>’s <code>xlim</code>
and <code>ylim</code> functions.</p>
</div>
</div>
<div class="section level2">
<h2 id="final-product">Final product<a class="anchor" aria-label="anchor" href="#final-product"></a>
</h2>
<p>That’s about it for the most basic functionalities of the clonal
expansion visualization function. Remember to save the plot first as an
<code>.svg</code> file for maximal resolution, and it should be
publication ready.</p>
<p>For fun, here’s a side-by-side comparison of the UMAP plot and the
clonal expansion plot:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/" class="external-link">cowplot</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span></span>
<span>  <span class="va">umap_plot</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"scRNA-seq UMAP"</span><span class="op">)</span>,</span>
<span>  <span class="va">pbmc_expansion_plot</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"APackOfTheClones clonal expansion plot"</span><span class="op">)</span>,</span>
<span>  labels <span class="op">=</span> <span class="st">"AUTO"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="v0-main_files/figure-html/final-1.png" width="700"></p>
<details><summary><strong>How do I reproduce the plots in this vignette?</strong>
</summary><p>If you really want to reproduce the plots and code in this vignette,
you will need to install the data used in this vignette which has been
compiled in an R package <code>SCIPData</code> (Single Cell Immune
Profiling Data). You can do so with the following code</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install the data package</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://devtools.r-lib.org/" class="external-link">devtools</a></span><span class="op">)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"Qile0317/SCIPData"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># load the package data directly into R objects in memory</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"pbmc"</span>, <span class="st">"tcr_dataframe"</span><span class="op">)</span></span></code></pre></div>
<p>And the rest of the code can be copied and used accordingly.</p>
Alternatively, there are two smaller, minimal, artificial versions of
the seurat object and clonotype data that can be loaded in
<code>APackOfTheClones</code>, run <code><a href="https://qile0317.github.io/APackOfTheClones/reference/mini_seurat_obj.html">?mini_seurat_obj</a></code> and
<code><a href="https://qile0317.github.io/APackOfTheClones/reference/mini_clonotype_data.html">?mini_clonotype_data</a></code> for more details.
</details>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Qile Yang.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
