<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="A brief walkthrough of the clonal expansion visualization workflow.
">
<title>APackOfTheClones Essentials • APackOfTheClones</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="APackOfTheClones Essentials">
<meta property="og:description" content="A brief walkthrough of the clonal expansion visualization workflow.
">
<meta property="og:image" content="https://qile0317.github.io/APackOfTheClones/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-TSDH5XEN3S"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-TSDH5XEN3S');
</script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">APackOfTheClones</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.2.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/APackOfTheClones-install.html">Install</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../articles/APackOfTheClones.html">Get started</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-vignettes">Vignettes</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-vignettes">
    <a class="dropdown-item" href="../articles/APackOfTheClones-runs.html">Storing and Fine-Tuning APackOfTheClones Runs</a>
    <a class="dropdown-item" href="../articles/APackOfTheClones-shared.html">Working With Clonotypes Shared Across Clusters</a>
    <a class="dropdown-item" href="../articles/APackOfTheClones-utils.html">Clonal Analysis Utilities in APackOfTheClones</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-archive">Archive</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-archive">
    <a class="dropdown-item" href="../articles/archive/v0-main.html">v0.1.x main vignette</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/Qile0317/APackOfTheClones/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>APackOfTheClones Essentials</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/Qile0317/APackOfTheClones/blob/HEAD/vignettes/APackOfTheClones.Rmd" class="external-link"><code>vignettes/APackOfTheClones.Rmd</code></a></small>
      <div class="d-none name"><code>APackOfTheClones.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Single-cell RNA sequencing (scRNA-seq) and T/B cell receptor (TCR)
sequencing are popular techniques for studying immune cell function and
disease. The combined use of such data can provide valuable insights
into the immune system, including clonal expansion.
<code>APackOfTheClones</code> provides a simple, easily customizable,
and publication-ready method to intuitively visualize clonal expansion
between different cell clusters with <code>ggplot2</code>, and can be
easily slotted into any analysis pipeline.</p>
<p>The method counts clonotypes, and using a dimensional reduction
(e.g. UMAP) of all cells in a single cell immune profiling experiment as
a base, it circle-packs all clone sizes directly as circles within
circular clusters according to its seurat cluster. The advantage of this
method compared to other (very limited) visualizations of clonal
expansion out there is that its arguably much more intuitive. To see it
in action, see <a href="https://doi.org/10.1126/sciimmunol.abg6356" class="external-link">this
paper</a> and <a href="https://doi.org/10.1111/all.15399" class="external-link">this
paper</a>.</p>
<p>Basic familiarity with the <code>R</code> language, the
<code>Seurat</code> package, and the <code>scRepertoire</code> package
<em>VERSION TWO</em> is assumed. As version two of
<code>scRepertoire</code> is relatively new and introduces some small
breaking changes, it is <em>essential</em> that its <a href="https://www.borch.dev/uploads/screpertoire/" class="external-link">new vignettes</a> are
read.</p>
<p>In this vignette, the most essential functionalities of the package
are covered, from preparing the data with scRepertoire, to producing and
fine-tuning a barebones visualization of clonal expansion.</p>
<div class="section level3">
<h3 id="setting-up-the-seurat-object-and-receptor-library-with-screpertoire">Setting up the Seurat Object and Receptor Library with
scRepertoire<a class="anchor" aria-label="anchor" href="#setting-up-the-seurat-object-and-receptor-library-with-screpertoire"></a>
</h3>
<p>The premise of the package is that it provides an additional analysis
tool on top of <code>scRepertoire</code>’s many functions, which are all
ran after combining all TCR/BCR contigs into clones, and integrating the
clonal information into a seurat object with a dimensional reduction(s).
Here are the corresponding vignettes to read in order for combining
contigs:</p>
<ol style="list-style-type: decimal">
<li><a href="https://www.borch.dev/uploads/screpertoire/articles/loading" class="external-link">Loading
Data</a></li>
<li><a href="https://www.borch.dev/uploads/screpertoire/articles/combining_contigs" class="external-link">Combining
Contigs into Clones</a></li>
<li><a href="https://www.borch.dev/uploads/screpertoire/articles/processing" class="external-link">Additional
Processing Steps (optional)</a></li>
</ol>
<p>And here is a practical example on how to generate a combined seurat
+ VDJ object to be used in the vignette named <code>pbmc</code>, which
is also identical to the built-in example dataset
<code>combined_pbmc</code> which can be loaded with
<code>data("combined_pbmc", package = "APackOfTheClones")</code>:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scRepertoire</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># load in the corresponding 6-sample TCR contigs from scRepertoire</span></span>
<span><span class="va">contig_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/get.html" class="external-link">get</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"contig_list"</span>, package <span class="op">=</span> <span class="st">"scRepertoire"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># combine the TCR contigs into clones with custom samples</span></span>
<span><span class="va">combined_contig_list</span> <span class="op">&lt;-</span> <span class="fu">scRepertoire</span><span class="fu">::</span><span class="fu">combineTCR</span><span class="op">(</span></span>
<span>  <span class="va">contig_list</span>,</span>
<span>  samples <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"P17B"</span>, <span class="st">"P17L"</span>, <span class="st">"P18B"</span>, <span class="st">"P18L"</span>, <span class="st">"P19B"</span>, <span class="st">"P19L"</span>, <span class="st">"P20B"</span>, <span class="st">"P20L"</span><span class="op">)</span>,</span>
<span>  removeNA <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>  removeMulti <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>  filterMulti <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>For integrating the clonal information,
<code>scRepertoire::combineExpression</code> is the function used <a href="https://www.borch.dev/uploads/screpertoire/articles/attaching_sc" class="external-link">to
do so</a>, but for <code>APackOfTheClones</code> the scRNA-seq object
has to be a seurat object:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># a seurat object corresponding to combined_contig_list named `pbmc` is loaded</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu">scRepertoire</span><span class="fu">::</span><span class="fu">combineExpression</span><span class="op">(</span></span>
<span>  <span class="va">combined_contig_list</span>,</span>
<span>  <span class="va">pbmc</span>,</span>
<span>  cloneCall <span class="op">=</span> <span class="st">"gene"</span>,</span>
<span>  proportion <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>And as a reference, here is the corresponding UMAP plot of the seurat
object</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">pbmc_umap_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">UMAPPlot</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span>
<span><span class="va">pbmc_umap_plot</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="APackOfTheClones_files/figure-html/umap-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="simple-ball-packing-visualization">Simple ball packing visualization<a class="anchor" aria-label="anchor" href="#simple-ball-packing-visualization"></a>
</h2>
<p><code>vizAPOTC</code> (short for “visualize APackOfTheClones”) is the
main convenience function of the package to directly produce the ball
packing clonal expansion plot. It takes in a main argument of a combined
seurat object (with a long list of optional arguments which will be
covered in later sections), and outputs a ggplot object:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">default_apotc_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/vizAPOTC.html">vizAPOTC</a></span><span class="op">(</span><span class="va">pbmc</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">default_apotc_plot</span></span></code></pre></div>
<p><img src="APackOfTheClones_files/figure-html/initial_vizapotc-1.png" width="700"></p>
<div class="section level3">
<h3 id="key-characteristics-of-the-plot">Key characteristics of the plot<a class="anchor" aria-label="anchor" href="#key-characteristics-of-the-plot"></a>
</h3>
<ul>
<li>One should notice the immediate correspondence of circular clusters
to umap clusters, and be able to derive some immediate insights about
which clusters have expanded.</li>
<li>The clonotype counts for each seurat cluster corresponds to their
position and color in the original UMAP centroids</li>
<li>The most expanded clonotypes are in the center of each circle
cluster with larger sizes, symbolizing increased expansion</li>
<li>On the plot, there is a somewhat imperfect visual legend of the
relative clone sizes</li>
<li>Some clusters have considerable visual overlap</li>
<li>The returned plot is a fully customizable <code>ggplot</code>
object</li>
</ul>
<p>The resulting clonal expansion plot may not be visually satisfactory
on the first run without customizations from optional arguments. These
arguments and other <code>ggplot</code> tricks will be covered to
fine-tune the visualization until publication-ready.</p>
</div>
</div>
<div class="section level2">
<h2 id="working-with-different-reduction-bases-and-clonal-representations">Working with different reduction bases and clonal
representations<a class="anchor" aria-label="anchor" href="#working-with-different-reduction-bases-and-clonal-representations"></a>
</h2>
<p>The arguments to do so are:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a>reduction_base <span class="ot">=</span> <span class="st">"umap"</span>,</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a>clonecall <span class="ot">=</span> <span class="st">"strict"</span>,</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a>alt_ident <span class="ot">=</span> <span class="cn">NULL</span></span></code></pre></div>
<p><code>reduction_base</code> indicates what each seurat cluster’s
centroid locations should be based upon, so UMAP, T-SNE, or PCA
(provided the reduction has been ran on the object).
<code>clonecall</code> corresponds to which column in the seurat
metadata should be used to conduct the clonotype counting. These are
usually columns generated by combineTCR/BCR but can be custom columns
too. Note that <code>scRepertoire</code> should generate the following
columns by default, and the brackets indicate equivalent names one can
pass into the clonecall parameter:</p>
<ul>
<li>CTgene (gene, genes)</li>
<li>CTnt (nt, nucleotide, nucleotides)</li>
<li>CTaa (aa, amino)</li>
<li>CTstrict (strict, gene+nt)</li>
</ul>
<p>Switching this parameter could be useful if the user had generated a
custom definition of clones and added it to the meta data, or if BCR
clonal data is present, since <a href="https://doi.org/10.3389/fimmu.2023.1123968" class="external-link">the definition of a
clone for B-cells isn’t as clear as it is for TCRs</a>.</p>
<p>Lastly, by default, clonotypes are grouped into clusters based on the
current active identity of all cells (<code>Idents(pbmc)</code>) which
is the case when <code>alt_ident = NULL</code>. However, to work with
some alternative ident that was added to the seurat metadata either
manually or via <code><a href="https://rdrr.io/pkg/SeuratObject/man/Idents.html" class="external-link">Seurat::StashIdent</a></code>, simply set
<code>alt_ident</code> to a character representing the name of that
column.</p>
</div>
<div class="section level2">
<h2 id="working-with-a-data-subset">Working With a Data Subset<a class="anchor" aria-label="anchor" href="#working-with-a-data-subset"></a>
</h2>
<p>A novel feature of version 1 is the ease of running APackOfTheClones
for subsets of the full dataset, which may be useful in cases like the
need for clonal expansion plotting for only certain samples or
conditions. There are two arguments to do so:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a>...,</span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>extra_filter <span class="ot">=</span> <span class="cn">NULL</span></span></code></pre></div>
<p><code>...</code> represents an arbitrary number of additional keyword
arguments indicating the rows corresponding to elements in the seurat
object metadata that should be filtered by. For example, seurat_clusters
= c(1, 9, 10) will filter the cells to only those in seurat clusters 1,
9, and 10. <code>extra_filter</code> is another additional way to subset
the data and should be formatted <em>exactly</em> like a statement one
would pass into <code><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">dplyr::filter</a></code> that does <em>additional</em>
filtering to cells in the seurat object.</p>
<p>Here is an example of the data subsetting, where only samples in the
example seurat object that correspond to sample 17, 18, and 19 are
plotted, alongside the original plot to see the difference. Optionally,
it may sometimes be helpful to set one of the visual parameters
<code>retain_axis_scales</code> to <code>TRUE</code>, extending the
output plot’s dimensions if needed to approximately match the
corresponding reduction plot’s dimensions:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># `orig.ident` is a custom column in the example data with levels corresponding to sample ids:</span></span>
<span><span class="co"># ("P17B" "P17L" "P18B" "P18L" "P19B" "P19L" "P20B" "P20L"). Here, it is subsetted</span></span>
<span><span class="co"># by the keyword argument approach</span></span>
<span><span class="va">subset_sample_17_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/vizAPOTC.html">vizAPOTC</a></span><span class="op">(</span></span>
<span>  <span class="va">pbmc</span>,</span>
<span>  orig.ident <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"P17B"</span>, <span class="st">"P17L"</span><span class="op">)</span>,</span>
<span>  retain_axis_scales <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  add_size_legend <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># here, it is subsetted with `extra_filter` for sample 18 with dplyr syntax:</span></span>
<span><span class="va">subset_sample_18_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/vizAPOTC.html">vizAPOTC</a></span><span class="op">(</span></span>
<span>  <span class="va">pbmc</span>,</span>
<span>  extra_filter <span class="op">=</span> <span class="st">"substr(orig.ident, 1, 3) == 'P18'"</span>,</span>
<span>  retain_axis_scales <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  add_size_legend <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># here, sample 19 is subsetted with both arguments to show that they work in conjunction</span></span>
<span><span class="va">subset_sample_19_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/vizAPOTC.html">vizAPOTC</a></span><span class="op">(</span></span>
<span>  <span class="va">pbmc</span>,</span>
<span>  orig.ident <span class="op">=</span> <span class="st">"P19B"</span>,</span>
<span>  extra_filter <span class="op">=</span> <span class="st">"orig.ident == 'P19L' | orig.ident == 'P19B'"</span>,</span>
<span>  retain_axis_scales <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  add_size_legend <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="../reference/vizAPOTC.html">vizAPOTC</a></span><span class="op">(</span><span class="va">combined_pbmc</span>, add_size_legend <span class="op">=</span> <span class="cn">FALSE</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>,</span>
<span>  <span class="va">subset_sample_17_plot</span>,</span>
<span>  <span class="va">subset_sample_18_plot</span>,</span>
<span>  <span class="va">subset_sample_19_plot</span>,</span>
<span>  labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"all"</span>, <span class="st">"17"</span>, <span class="st">"18"</span>, <span class="st">"19"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="APackOfTheClones_files/figure-html/subsetting-1.png" width="576"></p>
<div class="section level3">
<h3 id="a-note-on-subsetting-data">A Note on Subsetting Data<a class="anchor" aria-label="anchor" href="#a-note-on-subsetting-data"></a>
</h3>
<p>The subsetting parameters are barely more than syntactic sugar to
subset the seurat metadata and dimensional reduction coordinates, and
serves as a convenience to the user if they want to quickly generate
insights. Generally, if the analysis of data subsets (such as certain
samples, or condition, etc.) is crucial to a publication, one would
likely already be working with fully subsetted seurat objects, and
APackOfTheClones functions can be directly applied in those cases. See
the corresponding Seurat <a href="https://satijalab.org/seurat/articles/essential_commands.html#subsetting-and-merging" class="external-link"><code>vignette("essential_commands")</code></a>
for details.</p>
</div>
</div>
<div class="section level2">
<h2 id="customization-of-visual-parameters">Customization of Visual Parameters<a class="anchor" aria-label="anchor" href="#customization-of-visual-parameters"></a>
</h2>
<div class="section level3">
<h3 id="visually-scaling-circle-sizes">Visually Scaling Circle Sizes<a class="anchor" aria-label="anchor" href="#visually-scaling-circle-sizes"></a>
</h3>
<p>For each seurat cluster at index <span class="math inline">\(i\)</span>, the final radii <span class="math inline">\(r_{ij}\)</span> of each physical circle at index
<span class="math inline">\(j\)</span> representing clone size <span class="math inline">\(s_{ij}\)</span> is calculated with a clone size
scaling factor <span class="math inline">\(C \in (0, \infty)\)</span>
and a radius scaling factor <span class="math inline">\(R \in (0,
1]\)</span> with the following formula:</p>
<p><span class="math display">\[
r_{ij} = C \cdot (\sqrt{s_{ij}} - (1 - R))
\]</span></p>
<p>Intuitively, <span class="math inline">\(C\)</span> represents how
much to enlarge/shrink each radius geometrically relative to the clone
size, whereas <span class="math inline">\(R\)</span> represents the
scaled size of the smallest clone’s radius, and all circles will have
their radii decreased by that amount. <span class="math inline">\(R\)</span> defaults to 0.95, whereas <span class="math inline">\(C\)</span> defaults to an approximated factor
based on the number of clones that the package computes. The
corresponding arguments are <code>clone_scale_factor</code> and
<code>rad_scale_factor</code> in <code>vizAPOTC</code>.</p>
</div>
<div class="section level3">
<h3 id="cluster-repulsion">Cluster Repulsion<a class="anchor" aria-label="anchor" href="#cluster-repulsion"></a>
</h3>
<p>Considerable visual overlap between clusters (due to the algorithm’s
attempt to fit clusters to the original UMAP coordinates) may occur and
obstruct eachother excessively. APackOfTheClones by default tries to
alleviate this by repulsing overlapping clusters away from eachother.
There are four optional arguments:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a>repulse <span class="ot">=</span> <span class="cn">TRUE</span>,</span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a>repulsion_threshold <span class="ot">=</span> <span class="dv">1</span>,</span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a>repulsion_strength <span class="ot">=</span> <span class="dv">1</span>,</span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a>max_repulsion_iter <span class="ot">=</span> <span class="dv">10</span></span></code></pre></div>
<ul>
<li>
<code>repulsion_threshold</code> indicates the amount of
<code>ggplot2</code> units of overlap between clusters that are
acceptable. It defaults to <code>1</code>, meaning that two clusters
that overlap by about 1 unit are considered by the repulsion algorithm
to not be overlapping. Increasing this number will increase the amount
of overlap between clusters, and decreasing this number will do the
opposite, while decreasing it to negative values will incur additional
spacing between clusters. Note that as of the current version, two
clusters are considered overlapping by comparing their approximated
radii, which is computed from the right-most x coordinate. This isn’t a
very accurate approximation for smaller clusters with non-circular
borders.</li>
<li>
<code>repulsion_strength</code> relates to how much the clusters
should repel each other. The repulsion algorithm works in iterations,
where for each iteration, each cluster “pushes” each other away from
eachother by some amount. Increasing this value will cause extra
“pushing” during each iteration. However, increasing this factor too
much may result once again in a very visually unpleasant plot.</li>
<li>
<code>max_repulsion_iter</code> indicates the number of iterations
where clusters should repel eachother. Increasing this number would
ensure that clusters will (almost always) for sure not be overlapping. A
trick with this parameter to make more pleasant plots is to decrease
<code>repulsion_strength</code> and increase
<code>max_repulsion_iter</code> to possibly make a more pleasant
arrangement of clusters.</li>
</ul>
<p>For more details on them, read the “Arguments” section in the
function documentation. Briefly, first, to make the circle clusters move
away from eachother, <code>repulse</code> should be set to
<code>TRUE</code>, and the function should be ran AGAIN. (If you feel
this excessive re-running takes too long or is inefficient for your
workflow in its current form due to a constant need to fine-tune, see
the vignette <code>APackOfTheClones-runs</code>.)</p>
</div>
<div class="section level3">
<h3 id="managing-the-clone-size-legend">Managing the Clone Size Legend<a class="anchor" aria-label="anchor" href="#managing-the-clone-size-legend"></a>
</h3>
<p>There are the following six parameters to adjust the legend:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a>add_size_legend <span class="ot">=</span> <span class="cn">TRUE</span>,</span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a>legend_sizes <span class="ot">=</span> <span class="st">"auto"</span>,</span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a>legend_position <span class="ot">=</span> <span class="st">"auto"</span>,</span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a>legend_buffer <span class="ot">=</span> <span class="fl">0.2</span>,</span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a>legend_color <span class="ot">=</span> <span class="st">"#808080"</span>,</span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a>legend_spacing <span class="ot">=</span> <span class="st">"auto"</span>,</span>
<span id="cb9-7"><a href="#cb9-7" tabindex="-1"></a>legend_label <span class="ot">=</span> <span class="st">"Clone sizes"</span>,</span>
<span id="cb9-8"><a href="#cb9-8" tabindex="-1"></a>legend_text_size <span class="ot">=</span> <span class="dv">5</span>,</span>
<span id="cb9-9"><a href="#cb9-9" tabindex="-1"></a>add_legend_background <span class="ot">=</span> <span class="cn">TRUE</span>,</span></code></pre></div>
<p><code>legend_sizes</code> are autogenerated, with at least sizes for
clone size 1 and max(clone sizes). The user can also input whichever
sizes they wish as a numeric vector. The <code>legend_position</code> on
the plot defaults to one of the four corners, whichever will result in
the least overlap of the legend with circles present. Otherwise, the
user can input manually <code>"top left"</code>,
<code>"top right"</code>, <code>"bottom left"</code>
<code>"bottom right</code>, or just a numeric indicating the x and y
coordinates of the top leftmost corner of the legend.
<code>add_legend_background</code> will display a rectangular background
to the legend. The other parameters are less relevant and more details
about these arguments can be read in the function documentation.</p>
<p>Note that the functions <code>overlayLegend</code> and
<code>removeLegend</code> also exist for adding/removing legends on a
readily generated plot.</p>
</div>
<div class="section level3">
<h3 id="other-modifications">Other Modifications<a class="anchor" aria-label="anchor" href="#other-modifications"></a>
</h3>
<p>A collection of other utility parameters are:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a>order_clones <span class="ot">=</span> <span class="cn">TRUE</span>,</span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a>try_place <span class="ot">=</span> <span class="cn">FALSE</span>,</span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a>res <span class="ot">=</span> <span class="dv">360</span><span class="dt">L</span>,</span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a>linetype <span class="ot">=</span> <span class="st">"blank"</span>,</span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a>use_default_theme <span class="ot">=</span> <span class="cn">TRUE</span>,</span>
<span id="cb10-6"><a href="#cb10-6" tabindex="-1"></a>retain_axis_scales <span class="ot">=</span> <span class="cn">FALSE</span>,</span>
<span id="cb10-7"><a href="#cb10-7" tabindex="-1"></a>show_labels <span class="ot">=</span> <span class="cn">FALSE</span>,</span>
<span id="cb10-8"><a href="#cb10-8" tabindex="-1"></a>label_size <span class="ot">=</span> <span class="dv">5</span>,</span></code></pre></div>
<p>The most important of which are 1. <code>show_labels</code> which
overlays the seurat cluster on the plot in the format <code>"Cx"</code>
in the center of each relevant cluster. (see
<code><a href="../articles/APackOfTheClones-runs.html">vignette("APackOfTheClones-runs")</a></code> on how one can modify
them) 2. <code>use_default_theme</code> which will produce a plot that’s
thematically very similar to the dimensional reduction plot, with axis
labels and numeric ticks on each axis. This may be helpful for visually
indicating certain plot parameters. However, for publications <a href="https://doi.org/10.1126/sciimmunol.abg6356" class="external-link">such as this one</a>,
as the numeric ticks in APackOfTheClones have no actual biologically
relevant meaning aside from approximately matching clusters, it may be
beneficial to set this argument to <code>FALSE</code> which will produce
a plot with just the circles. The following is an example parameters to
generate a plot that is likely publication-ready:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/vizAPOTC.html">vizAPOTC</a></span><span class="op">(</span></span>
<span>  <span class="va">pbmc</span>,</span>
<span>  repulsion_threshold <span class="op">=</span> <span class="fl">0.9</span>,</span>
<span>  show_labels <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  label_size <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  legend_text_size <span class="op">=</span> <span class="fl">3.5</span>,</span>
<span>  add_legend_background <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  use_default_theme <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="APackOfTheClones_files/figure-html/void_labelled_plot-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="conclusion">Conclusion<a class="anchor" aria-label="anchor" href="#conclusion"></a>
</h2>
<p>That’s about it for the most basic functionalities of the clonal
expansion visualization function. It’s <strong><em>strongly
recommended</em></strong> to save the plot first as an <code>.svg</code>
file with <code><a href="https://ggplot2.tidyverse.org/reference/ggsave.html" class="external-link">ggplot2::ggsave</a></code> for maximal resolution
(especially of the circles) for publication.</p>
<p>For users that need to fine-tune plot parameters and/or save the data
within the seurat object for readjustment/replotting, please read the
<code><a href="../articles/APackOfTheClones-runs.html">vignette("APackOfTheClones-runs")</a></code>.</p>
<p>For inspiration of how it could practically look like in a real paper
context, see the following papers where the original julia
implementation of APackOfTheClones was successfully used:</p>
<ul>
<li><a href="https://doi.org/10.1126/sciimmunol.abg6356" class="external-link">Single-cell
analysis pinpoints distinct populations of cytotoxic CD4+ T cells and an
IL-10+CD109+ <span class="math inline">\(T_{H}2\)</span> cell population
in nasal polyps</a></li>
<li><a href="https://doi.org/10.1111/all.15399" class="external-link">Recombinant multimeric
dog allergen prevents airway hyperresponsiveness in a model of asthma
marked by vigorous <span class="math inline">\(T_{H}2\)</span> and <span class="math inline">\(T_{H}17\)</span> cell responses</a></li>
</ul>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Qile Yang.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
